internal/arduino/builder/internal/runner/runner.go

```diff
@@ -0,0 +1,112 @@
+// This file is part of arduino-cli.
+//
+// Copyright 2024 ARDUINO SA (http://www.arduino.cc/)
+//
+// This software is released under the GNU General Public License version 3,
+// which covers the main part of arduino-cli.
+// The terms of this license can be found at:
+// https://www.gnu.org/licenses/gpl-3.0.en.html
+//
+// You can be released from the requirements of the above licenses by purchasing
+// a commercial license. Buying such a license is mandatory if you want to
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package runner
+
+import (
+	"context"
+	"runtime"
+	"sync"
+)
+
+// Runner is a helper to run commands in a queue, the commands are immediately exectuded
+// in a goroutine as they are enqueued. The runner can be stopped by calling Cancel.
+type Runner struct {
+	lock      sync.Mutex
+	queue     chan<- *enqueuedCommand
+	results   map[string]<-chan *Result
+	ctx       context.Context
+	ctxCancel func()
+	wg        sync.WaitGroup
+}
+
+type enqueuedCommand struct {
+	task   *Task
+	accept func(*Result)
+}
+
+func (cmd *enqueuedCommand) String() string {
+	return cmd.task.String()
+}
+
+func New(inCtx context.Context) *Runner {
+	ctx, cancel := context.WithCancel(inCtx)
+	queue := make(chan *enqueuedCommand, 1000)
+	r := &Runner{
+		ctx:       ctx,
+		ctxCancel: cancel,
+		queue:     queue,
+		results:   map[string]<-chan *Result{},
+	}
+
+	// Spawn workers
+	for i := 0; i < runtime.NumCPU(); i++ {
```

Comment:
I'd make this optional as we already did in: https://github.com/arduino/arduino-cli/blob/master/internal/arduino/builder/compilation.go#L78

We can reuse the `jobs` variable and pass it in the `SketchLibrariesDetector` so we can inject it in the runner constructor func.